ocaml/xapi/xapi_host.ml

```diff
@@ -2671,44 +2671,165 @@ let with_temp_file_contents ~contents f =
+      else
+        false
+    in
+    if not (is_valid_symlink from_path to_path) then (
+      try
+        Helpers.rmtree from_path ;
+        Unix.symlink to_path from_path
+      with e ->
+        let msg = Printf.sprintf "Failed to symlink %s->%s" from_path to_path in
+        error "%s: %s" msg (Printexc.to_string e) ;
+        raise Api_errors.(Server_error (internal_error, [msg]))
+    ) ;
+    fn from_path
+  in
+  let with_empty_dir path fn =
+    debug "override_uefi_certs: with_empty_dir %s" path ;
+    if Sys.file_exists path then (
+      Unix.(
+        match (lstat path).st_kind with
+        | S_DIR ->
+            Sys.readdir path
+            |> Array.iter (fun file -> Helpers.rmtree (path // file))
+        | _ ->
+            Unix.unlink path ;
+            Unixext.mkdir_rec path 0o755
+      )
+    ) else
+      Unixext.mkdir_rec path 0o755 ;
+    fn path
+  in
+  let check_valid_uefi_certs_in path =
+    (* check expected uefi certificates are present *)
+    let rec with_uefi_certs_in_disk path =
+      if Sys.file_exists path then
+        Unix.(
+          match (lstat path).st_kind with
+          | S_DIR ->
+              Sys.readdir path |> Array.to_list
+          | S_LNK ->
+              Unix.readlink path |> with_uefi_certs_in_disk
+          | _ ->
+              []
+        )
+      else
+        []
+    in
+    let uefi_certs_in_disk = with_uefi_certs_in_disk path in
+    ["KEK.auth"; "db.auth"]
+    |> List.iter (fun cert ->
+           let log_of found =
+             (if found then debug else error)
+               "check_valid_uefi_certs: %s %s in %s"
+               (if found then "found" else "missing")
+               cert path
+           in
+           log_of (List.mem cert uefi_certs_in_disk)
+       )
+  in
+  match !Xapi_globs.override_uefi_certs with
+  | false ->
+      with_valid_symlink !Xapi_globs.varstore_dir !Xapi_globs.default_auth_dir
+        (fun path -> check_valid_uefi_certs_in path
+      )
+  | true ->
+      with_empty_dir !Xapi_globs.varstore_dir (fun path ->
+          (* get from pool for consistent results across hosts *)
+          let pool_uefi_certs =
+            Db.Pool.get_uefi_certificates ~__context
+              ~self:(Helpers.get_pool ~__context)
+          in
+          really_write_uefi_certificates_to_disk ~__context ~host
+            ~value:pool_uefi_certs ;
+          check_valid_uefi_certs_in path
+      )
 
 let set_uefi_certificates ~__context ~host ~value =
-  Db.Host.set_uefi_certificates ~__context ~self:host ~value ;
-  Helpers.call_api_functions ~__context (fun rpc session_id ->
-      Client.Client.Pool.set_uefi_certificates ~rpc ~session_id
-        ~self:(Helpers.get_pool ~__context)
-        ~value
-  )
+  match !Xapi_globs.override_uefi_certs with
+  | false ->
+      raise
+        Api_errors.(
+          Server_error
+            ( Api_errors.operation_not_allowed
+            , ["override_uefi_certs not enabled"]
```

Comment:
I would rather leave the message empty, this action might not be supported, depending on the product used. How to enable this can be instead be explained in the docs of xcp-ng since it's a special case for customers already using custom certificates, which Windows guests might not consider kosher

Reply:
I believe I fixed it in the last patchset in the way your expected, I wonder if you could double-check

Reply:
It looks good now, thanks